common/determined_common/_logging.py

```diff
@@ -1,16 +1,207 @@
+                harness_log_level="DEBUG",
+                debug_all_workers=True,
+                horovod_verbose=True,
+                nccl_debug=None,
+                nccl_debug_subsys=None,
+                resource_profile_period_sec=10.0,
+                stack_trace_period_sec=0.0,
+            )
+
+        assert isinstance(config, dict)
+        return cls(**config)
+
+    @classmethod
+    def from_environ(cls, environ: Dict[str, str]) -> "DebugConfig":
+        if environ.get("DET_DEBUG"):
+            # If DET_DEBUG is set to a non-empty string, behave like `debug: true` in the config.
+            return cls.from_config(True)
+        return cls()
+
+    def set_workload(self, workload: str) -> None:
+        """Update the workload information to associate with debug messages."""
+        self._workload = workload
+
+    @contextlib.contextmanager
+    def profiling_thread(self, machine_rank: int, worker_rank: int) -> Generator:
+        """If enabled, emit profiling information periodically."""
+        if self._resource_profile_period_sec <= 0.0:
+            yield
+            return
+
+        def profile_fn() -> None:
+            msg = {
+                "time": time.time(),
+                "machine_rank": machine_rank,
+                "worker_rank": worker_rank,
+                "workload": self._workload,
+            }
+            log.resources.debug(json.dumps(msg))
+
+        thread = _ProfilingThread(self._resource_profile_period_sec, profile_fn)
+        try:
+            yield
+        finally:
+            thread.quit()
+            thread.join()
+
+    @contextlib.contextmanager
+    def stack_trace_thread(self) -> Generator:
+        """If enabled, emit stack traces periodically."""
+        if self._stack_trace_period_sec <= 0.0:
+            yield
+            return
+
+        import faulthandler
+
+        faulthandler.dump_traceback_later(self._stack_trace_period_sec, repeat=True)
+        try:
+            yield
+        finally:
+            faulthandler.cancel_dump_traceback_later()
+
+    @property
+    def resource_profile_period_sec(self) -> float:
+        return self._resource_profile_period_sec
+
+    @property
+    def nccl_debug(self) -> Optional[str]:
+        return self._nccl_debug
+
+    @property
+    def nccl_debug_subsys(self) -> Optional[str]:
+        return self._nccl_debug_subsys
+
+    @property
+    def horovod_verbose(self) -> bool:
+        return self._horovod_verbose
+
+    @property
+    def debug_all_workers(self) -> bool:
+        return self._debug_all_workers
+
+    def set_loggers(
+        self, is_chief: bool = True, handler: Optional[logging.Handler] = None
+    ) -> None:
+        if not handler:
+            handler = logging.StreamHandler(sys.stdout)
+        formatter = logging.Formatter("%(levelname)s:%(name)s: %(message)s")
+        handler.setFormatter(formatter)
+
+        # Configure the root log handler for external libraries.
```

Comment:
Yes since logging settings are hierarchical.  If pytorch uses `logging.info()` or `logging.getLogger("torch").info()`, they will both be affected (although if you did `logging.getLogger("torch").setLevel(...)` you could override the root logger settings).